components/form/nz-form.directive.ts

```diff
@@ -76,6 +78,7 @@ export class NzFormDirective implements OnInit, OnChanges, AfterContentInit, OnD
   ngAfterContentInit(): void {
     this.nzFormLabelComponent.changes.pipe(startWith(null), takeUntil(this.destroy$)).subscribe(() => {
       this.updateItemsDefaultColon();
+      this.cdr.detectChanges();
```

Comment:
Please use `markForCheck` instead.

Reply:
`detectChanges` will trigger ancestor components' change detection once again, so using `markForCheck` function to mark this component as dirty will be better implementation? Do I get your point?

Reply:
`updateItemsDefaultColon` will be executed after the child component `nzFormLabelComponent` changes are detected. You must use detectChanges to re-trigger the child component change detection, otherwise it will cause `ExpressionChangedAfterItHasBeenCheckedError`, `markForCheck` will not trigger child component change detection.

Reply:
I haven't used a demo to verify this code, but as you said, 

> `updateItemsDefaultColon` function will be executed after the child component `nzFormLabelComponent ` changes are detected

Function `this.updateItemsDefaultColon()`  will call function `setDefaultNoColon()` iteratively, then `setDefalutNoColon` will mark `NzFormLabelComponent` as dirty by calling `this.cdr.markForCheck();` which will trigger change detecting for `NzFormLabelComponent`.

So, as far as I know, there should be no `ExpressionChangedAfterItHasBeenCheckedError ` here. 
Trying to create a demo for DB-checking.

Reply:
https://stackblitz.com/edit/angular-pztsyb?embed=1&file=src/app/app.component.ts

Reply:
`setDefaultNoColon` is directly assigned defaultNoColon without `@Input()`, so `NzFormLabelComponent` will not be marked as dirty, `this.cdr.markForCheck()` will only update `NzFormDirective`, and it is over.

Reply:
I see, even though `NzFormLabelComponent` & `formDirective` already has been marked as dirty, and will be detected in future point, but before next change detection cycle coming (we don't know the exact time), the old value does not equal to the current value, which causes the `ExpressionChangedAfterItHasBeenCheckedError `.Eventually, using `detectChanges` calling, at last, is the required  operation.

Reply:
> `setDefaultNoColon` is directly assigned `defaultNoColon` without `@Input()`, so `NzFormLabelComponent` will not be marked as dirty, `this.cdr.markForCheck()` will only update `NzFormDirective`, and it is over.

Not exactly, though without `@Input()`, every time function `setDefaultNoColon` from `NzFormLabelComponent`  being called, it is also calling `this.cdr.markForCheck();` statement for belonging `NzFormLabelComponent`, AKA , it will be marked as dirty. But  no matter what, `ExpressionChangedAfterItHasBeenCheckedError ` existing.